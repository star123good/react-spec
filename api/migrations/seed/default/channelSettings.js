const constants = require('./constants');
const { SPECTRUM_PRIVATE_CHANNEL_ID } = constants;

module.exports = [
  {
    id: 1,
    channelId: SPECTRUM_PRIVATE_CHANNEL_ID,
    joinSettings: {
      tokenJoinEnabled: true,
      token: 'abc',
    },
    slackSettings: {
      botLinks: {
        threadCreated: null,
      },
    },
  },
];
