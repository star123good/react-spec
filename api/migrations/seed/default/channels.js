// @flow
const constants = require('./constants');
const {
  DATE,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  PAYMENTS_GENERAL_CHANNEL_ID,
  PAYMENTS_PRIVATE_CHANNEL_ID,
  SPECTRUM_ARCHIVED_CHANNEL_ID,
  SPECTRUM_DELETED_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
  MODERATOR_CREATED_CHANNEL_ID,
  PRIVATE_GENERAL_CHANNEL_ID,
} = constants;

module.exports = [
  {
    id: SPECTRUM_GENERAL_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General chatter',
    slug: 'general',
    isPrivate: false,
    isDefault: true,
  },

  {
    id: SPECTRUM_PRIVATE_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Private',
    description: 'Private chatter',
    slug: 'private',
    isPrivate: true,
    isDefault: false,
  },

  {
    id: PAYMENTS_GENERAL_CHANNEL_ID,
    communityId: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General chatter',
    slug: 'general',
    isPrivate: false,
    isDefault: true,
  },

  {
    id: PAYMENTS_PRIVATE_CHANNEL_ID,
    communityId: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Private',
    description: 'Private chatter',
    slug: 'private',
    isPrivate: true,
    isDefault: false,
  },

  {
    id: SPECTRUM_ARCHIVED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Archived',
    description: 'Testing archiving',
    slug: 'archived',
    isPrivate: false,
    isDefault: true,
    archivedAt: new Date(DATE),
  },

  {
    id: SPECTRUM_DELETED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Deleted',
    description: 'Testing deleted channel',
    slug: 'deleted',
    isPrivate: false,
    isDefault: false,
    deletedAt: new Date(DATE),
  },

  {
    id: DELETED_COMMUNITY_DELETED_CHANNEL_ID,
    communityId: DELETED_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Deleted',
    description: 'Testing deleted channel',
    slug: 'deleted',
    isPrivate: false,
    isDefault: false,
    deletedAt: new Date(DATE),
  },

  {
    id: MODERATOR_CREATED_CHANNEL_ID,
    communityId: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'Moderator created',
    description: 'Moderator created channel',
    slug: 'moderator-created',
    isPrivate: false,
    isDefault: false,
  },

  {
    id: PRIVATE_GENERAL_CHANNEL_ID,
    communityId: PRIVATE_COMMUNITY_ID,
    createdAt: new Date(DATE),
    name: 'General',
    description: 'General',
    slug: 'private-general',
    isPrivate: false,
    isDefault: false,
  },
];
