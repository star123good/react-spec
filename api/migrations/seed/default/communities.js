// @flow
const constants = require('./constants');
const {
  DATE,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
} = constants;

module.exports = [
  {
    id: SPECTRUM_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: false,
    name: 'Spectrum',
    description: 'The future of communities',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'spectrum',
  },
  {
    id: PAYMENTS_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: false,
    name: 'Payments',
    description: 'Where payments are tested',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'payments',
  },
  {
    id: DELETED_COMMUNITY_ID,
    createdAt: new Date(DATE),
    deletedAt: new Date(DATE),
    isPrivate: false,
    name: 'Deleted',
    description: 'Things didnt work out',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'deleted',
  },
  {
    id: PRIVATE_COMMUNITY_ID,
    createdAt: new Date(DATE),
    isPrivate: true,
    name: 'Private community',
    description: 'Private community',
    website: 'https://spectrum.chat',
    profilePhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Profile.png.0.6225566835336693',
    coverPhoto:
      'https://spectrum.imgix.net/communities/-Kh6RfPYjmSaIWbkck8i/Twitter Header.png.0.3303118636071434',
    slug: 'private',
  },
];
