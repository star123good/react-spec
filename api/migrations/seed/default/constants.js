// @flow
const DATE = 1483225200000;

// users
const MAX_ID = '1';
const BRIAN_ID = '2';
const BRYN_ID = '3';
// this user is blocked in spectrum community
const BLOCKED_USER_ID = '4';
// this user is has never joined communities or channels
const QUIET_USER_ID = '5';
// this user was a previous member of spectrum community
const PREVIOUS_MEMBER_USER_ID = '6';
// this user is pending in all private channels
const PENDING_USER_ID = '7';
// this user is moderator in all channels, member in all communities
const CHANNEL_MODERATOR_USER_ID = '8';
// this user is moderator in all communities
const COMMUNITY_MODERATOR_USER_ID = '9';

// communities
const SPECTRUM_COMMUNITY_ID = '1';
const PAYMENTS_COMMUNITY_ID = '2';
const DELETED_COMMUNITY_ID = '3';
const PRIVATE_COMMUNITY_ID = '4';

// channels
const SPECTRUM_GENERAL_CHANNEL_ID = '1';
const SPECTRUM_PRIVATE_CHANNEL_ID = '2';
const PAYMENTS_GENERAL_CHANNEL_ID = '3';
const PAYMENTS_PRIVATE_CHANNEL_ID = '4';
const SPECTRUM_ARCHIVED_CHANNEL_ID = '5';
const SPECTRUM_DELETED_CHANNEL_ID = '6';
const DELETED_COMMUNITY_DELETED_CHANNEL_ID = '7';
const MODERATOR_CREATED_CHANNEL_ID = '8';
const PRIVATE_GENERAL_CHANNEL_ID = '9';

module.exports = {
  DATE,
  MAX_ID,
  BRIAN_ID,
  BRYN_ID,
  BLOCKED_USER_ID,
  QUIET_USER_ID,
  PREVIOUS_MEMBER_USER_ID,
  PENDING_USER_ID,
  CHANNEL_MODERATOR_USER_ID,
  COMMUNITY_MODERATOR_USER_ID,
  SPECTRUM_COMMUNITY_ID,
  PAYMENTS_COMMUNITY_ID,
  DELETED_COMMUNITY_ID,
  PRIVATE_COMMUNITY_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  PAYMENTS_GENERAL_CHANNEL_ID,
  PAYMENTS_PRIVATE_CHANNEL_ID,
  SPECTRUM_ARCHIVED_CHANNEL_ID,
  SPECTRUM_DELETED_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
  MODERATOR_CREATED_CHANNEL_ID,
  PRIVATE_GENERAL_CHANNEL_ID,
};
