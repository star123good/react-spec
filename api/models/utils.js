// @flow
import { db } from 'shared/db';

export type Timeframe = 'daily' | 'weekly' | 'monthly' | 'quarterly';

export const parseRange = (timeframe?: Timeframe) => {
  switch (timeframe) {
    case 'daily': {
      return { current: 60 * 60 * 24, previous: 60 * 60 * 24 * 2 };
    }
    case 'weekly': {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
    case 'monthly': {
      return { current: 60 * 60 * 24 * 30, previous: 60 * 60 * 24 * 60 };
    }
    case 'quarterly': {
      return { current: 60 * 60 * 24 * 90, previous: 60 * 60 * 24 * 180 };
    }
    default: {
      return { current: 60 * 60 * 24 * 7, previous: 60 * 60 * 24 * 14 };
    }
  }
};

export const getAu = (range: Timeframe) => {
  const { current } = parseRange(range);
  return db
    .table('users')
    .filter(row =>
      row
        .hasFields('lastSeen')
        .and(row('lastSeen').during(db.now().sub(current), db.now()))
    )
    .count()
    .default(0)
    .run();
};

// prettier-ignore
export const getGrowth = async (table: string, range: Timeframe, field: string, filter: ?mixed) => {
  const { current, previous } = parseRange(range);
  const currentPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(current), db.now()))
    .filter(filter ? filter : '')
    .count()
    .run();

  const prevPeriodCount = await db
    .table(table)
    .filter(db.row(field).during(db.now().sub(previous), db.now().sub(current)))
    .filter(filter ? filter : '')
    .count()
    .run();

  const rate = (await (currentPeriodCount - prevPeriodCount)) / prevPeriodCount;
  return {
    currentPeriodCount,
    prevPeriodCount,
    growth: Math.round(rate * 100),
  };
};

export const getCount = (table: string, filter: mixed) => {
  if (filter) {
    return db
      .table(table)
      .filter(filter)
      .count()
      .run();
  }

  return db
    .table(table)
    .count()
    .run();
};

export const getCoreMetrics = () => {
  return db
    .table('coreMetrics')
    .orderBy(db.desc('date'))
    .limit(90)
    .orderBy('date')
    .run();
};
