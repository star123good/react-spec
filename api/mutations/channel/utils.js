// @flow
import createLoaders from '../../loaders/';
const loaders = createLoaders();

export const userCanManageChannel = async (
  userId: string,
  channelId: string
): Promise<boolean> => {
  if (!userId || !channelId) return false;

  const channel = await loaders.channel.load(channelId);

  if (!channel || channel.deletedAt) return false;
  const [communityPermissions, channelPermissions] = await Promise.all([
    loaders.userPermissionsInCommunity.load([userId, channel.communityId]),
    loaders.userPermissionsInChannel.load([userId, channelId]),
  ]);

  if (!communityPermissions) return false;
  if (communityPermissions.isOwner || communityPermissions.isModerator)
    return true;
  if (!channelPermissions) return false;
  if (channelPermissions.isOwner || channelPermissions.isModerator) return true;
  return false;
};

export const userCanCreateChannel = async (
  userId: string,
  communityId: string
): Promise<boolean> => {
  if (!userId || !communityId) return false;
  const permissions = loaders.userPermissionsInCommunity.load([
    userId,
    communityId,
  ]);
  if (!permissions || !permissions.isOwner) return false;
  return true;
};
