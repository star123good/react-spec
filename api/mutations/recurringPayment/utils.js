// @flow
require('now-env');
import UserError from '../../utils/UserError';
import { stripe } from 'shared/stripe';

type Err = {
  type: string,
};

export const parseStripeErrors = (err: Err) => {
  switch (err.type) {
    case 'StripeCardError':
      // A declined card error
      return new UserError(err); // => e.g. "Your card's expiration year is invalid."
    case 'RateLimitError':
      // Too many requests made to the API too quickly
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 1'
      );
    case 'StripeInvalidRequestError':
      // Invalid parameters were supplied to Stripe's API
      return new UserError(
        'Could not upgrade to Pro at this time, try again later: 2'
      );
    case 'StripeAPIError':
      // An error occurred internally with Stripe's API
      return new UserError(
        'Something went wrong at Stripe, try again later: 3'
      );
    case 'StripeConnectionError':
      // Some kind of error occurred during the HTTPS communication
      return new UserError(
        'Something went wrong at Stripe, try again later: 4'
      );
    case 'StripeAuthenticationError':
      // You probably used an incorrect API key
      return new UserError(
        'Something went wrong at Stripe, try again later: 5'
      );
    default:
      // Handle any other types of unexpected errors
      return new UserError('Something went wrong, try again later: 6');
  }
};

export const getStripeCustomer = (customerId: string) => {
  try {
    return stripe.customers.retrieve(customerId);
  } catch (err) {
    return console.error(err) || err;
  }
};

export const createStripeCustomer = (email: string, source: string) => {
  try {
    return stripe.customers.create({
      email,
      source,
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const updateStripeCustomer = (
  customer: string,
  email: string,
  source: string
) => {
  try {
    return stripe.customers.update(customer, {
      email,
      source,
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const createStripeSubscription = (
  customer: string,
  plan: string,
  quantity: number
) => {
  try {
    return stripe.subscriptions.create({
      customer,
      items: [
        {
          plan,
          quantity: quantity ? quantity : 1, // if no quantity is provided, default to one
        },
      ],
    });
  } catch (err) {
    return console.error(err) || err;
  }
};

export const deleteStripeSubscription = (subscription: string) => {
  try {
    return stripe.subscriptions.del(subscription);
  } catch (err) {
    return console.error(err) || err;
  }
};
