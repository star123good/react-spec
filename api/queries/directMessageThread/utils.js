// @flow

export const canViewDMThread = async (
  threadId: string,
  userId: string,
  { loaders }: { loaders: Object }
) => {
  if (!userId) return false;

  const result = await loaders.directMessageParticipants.load(threadId);
  if (!result || !result.reduction || result.reduction.length === 0)
    return false;

  const members = result.reduction;
  const ids = members.map(({ userId }) => userId);
  if (ids.indexOf(userId) === -1) return false;

  return true;
};
