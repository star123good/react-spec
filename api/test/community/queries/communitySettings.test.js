//@flow
import { request } from '../../utils';
import { SPECTRUM_COMMUNITY_ID } from '../../../migrations/seed/default/constants';

it('should fetch a communitys settings', async () => {
  const query = /* GraphQL */ `
    {
      community(id: "${SPECTRUM_COMMUNITY_ID}") {
        id
      }
    }
  `;

  expect.assertions(2);
  const result = await request(query);
  expect(result).toMatchSnapshot();
});
