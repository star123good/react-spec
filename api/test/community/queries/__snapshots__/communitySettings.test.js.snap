// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch a communitys settings 1`] = `
Object {
  "data": Object {
    "community": Object {
      "id": "1",
    },
  },
}
`;
